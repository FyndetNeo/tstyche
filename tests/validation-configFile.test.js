import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, getFixtureUrl, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const fixtureUrl = getFixtureUrl("validation-configFile", { generated: true });

afterEach(async () => {
  await clearFixture(fixtureUrl);
});

describe("'tstyche.config.json' file", () => {
  test("handles unknown options", async () => {
    const config = {
      cache: "all",
      silent: true,
      testFileMatch: ["**/packages/*/__typetests__/*.test.ts"],
    };

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles option argument of wrong type", async () => {
    const config = {
      failFast: "always",
    };

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when tabs are used for indentation, handles option argument of wrong type", async () => {
    const configText = `{
\t"failFast": false,
\t"rootPath": true
}
`;

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: configText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles wrong root value", async () => {
    const config = [{ failFast: true }];

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles syntax error", async () => {
    const configText = `{
  'failFast': true
`;

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: configText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles single quoted property names", async () => {
    const configText = `{
  'failFast': true
}`;

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: configText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles single quoted values", async () => {
    const configText = `{
  "rootPath": '../'
}`;

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: configText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("handles single quoted list items", async () => {
    const configText = `{
  "target": ['4.8']
}`;

    await writeFixture(fixtureUrl, {
      ["tstyche.config.json"]: configText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(stdout).toBe("");
    expect(normalizeOutput(stderr)).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });
});

describe("'--config' command line option", () => {
  test("when option argument is missing", async () => {
    await writeFixture(fixtureUrl);

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--config"]);

    expect(stdout).toBe("");
    expect(stderr).toBe(
      [
        "Error: Option '--config' expects an argument.",
        "",
        "Option '--config' requires an argument of type string.",
        "",
        "",
      ].join("\n"),
    );

    expect(exitCode).toBe(1);
  });
});
