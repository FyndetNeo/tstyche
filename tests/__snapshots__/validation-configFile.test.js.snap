// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'tstyche.config.json' file handles list item of wrong type: stderr 1`] = `
"Error: Item of the 'testFileMatch' list must be of type string.

  1 | {
  2 |   "testFileMatch": [
> 3 |     true
    |     ^
  4 |   ]
  5 | }

      at ./tstyche.config.json:3:5

"
`;

exports[`'tstyche.config.json' file handles option argument of wrong type: stderr 1`] = `
"Error: Option 'failFast' requires an argument of type boolean.

  1 | {
> 2 |   "failFast": "always"
    |               ^
  3 | }

      at ./tstyche.config.json:2:15

"
`;

exports[`'tstyche.config.json' file handles single quoted list items: stderr 1`] = `
"Error: String literal with double quotes expected.

  1 | {
> 2 |   "target": ['4.8']
    |              ^
  3 | }

      at ./tstyche.config.json:2:14

"
`;

exports[`'tstyche.config.json' file handles single quoted property names: stderr 1`] = `
"Error: String literal with double quotes expected.

  1 | {
> 2 |   'failFast': true
    |   ^
  3 | }

      at ./tstyche.config.json:2:3

"
`;

exports[`'tstyche.config.json' file handles single quoted values: stderr 1`] = `
"Error: String literal with double quotes expected.

  1 | {
> 2 |   "rootPath": '../'
    |               ^
  3 | }

      at ./tstyche.config.json:2:15

"
`;

exports[`'tstyche.config.json' file handles syntax error: stderr 1`] = `
"Error: '}' expected. ts(1005)

  1 | {
  2 |   'failFast': true
> 3 | 
    | ^

      at ./tstyche.config.json:3:1

"
`;

exports[`'tstyche.config.json' file handles unknown options: stderr 1`] = `
"Error: Unknown option 'cache'.

  1 | {
> 2 |   "cache": "all",
    |   ^
  3 |   "silent": true,
  4 |   "testFileMatch": [
  5 |     "**/packages/*/__typetests__/*.test.ts"
  6 |   ]

      at ./tstyche.config.json:2:3

Error: Unknown option 'silent'.

  1 | {
  2 |   "cache": "all",
> 3 |   "silent": true,
    |   ^
  4 |   "testFileMatch": [
  5 |     "**/packages/*/__typetests__/*.test.ts"
  6 |   ]

      at ./tstyche.config.json:3:3

"
`;

exports[`'tstyche.config.json' file handles wrong root value: stderr 1`] = `
"Error: The root value of a configuration file must be an object literal.

> 1 | [
    | ^
  2 |   {
  3 |     "setupTimeout": 30
  4 |   }
  5 | ]

      at ./tstyche.config.json:1:1

"
`;

exports[`'tstyche.config.json' file when tabs are used for indentation, handles option argument of wrong type: stderr 1`] = `
"Error: Option 'rootPath' requires an argument of type string.

  1 | {
  2 |  "failFast": false,
> 3 |  "rootPath": true
    |              ^
  4 | }
  5 | 

      at ./tstyche.config.json:3:14

"
`;
